### DISCUSSION TITLE
Relacionamentos - EF

### DISCUSSION THREAD

**Discussion by marcus-silveira** [balta-io/forum#515] [Tira Dúvidas] [answered]
Eu só consigo rodar a aplicação se adicionar as PK ao Role e Tag:
modelBuilder.Entity<Role>().HasKey(x => x.Id);
modelBuilder.Entity<Tag>().HasKey(x=>x.Id);

Caso contrario dá o erro da imagem:
Unhandled exception. System.InvalidOperationException: The entity type 'Role' requires a primary key to be defined. If you intended to use a keyless entity type, call 'HasNoKey' in 'OnModelCreating'. For more information on keyless entity types, see https://go.microsoft.com/fwlink/?linkid=2141943.
![Captura de tela 2024-03-09 103352](https://github.com/balta-io/forum/assets/102529629/f496b4d4-bfeb-419a-9c2d-8f5ddcca48c9)

**Comment by henriqueweege** [comment, marked as answer]
Oi, @marcus-silveira! Tudo bem?
Você teria como postar aqui print das models das entidades?
Valeu!

**Reply by henriqueweege** [reply]
Bom, caso alguém esteja passando por algum problema parecido, da uma olhada [nessa doc da Microsoft](https://learn.microsoft.com/en-us/ef/ef6/modeling/code-first/conventions/built-in), e [nessa aqui também](https://learn.microsoft.com/en-us/ef/ef6/modeling/code-first/data-annotations). Basicamente o Entity não está conseguindo encontrar a chave primária. Usar o Annotation [Key] que está na segunda doc que coloquei aqui deve resolver!

Valeeeu!

**Reply by malighieri** [reply]
O @henriqueweege pediu pra você postar as classes que definem suas entidades pois o Entity Framework se baseia em nomenclaturas para inferir algumas coisas, incluindo chaves primárias e chaves estrangeiras. Por exemplo, em uma classe que contenha uma propriedade chamada **Id** o Entity Framework vai entender ela como sendo a chave primária. No caso de você ter uma relação entre duas entidades e precisa que seja gerada uma chave estrangeira, criando uma propriedade cujo nome seja [NomeDaEntidadeEstrangeira]+Id ele infere esta propriedade como sendo a chave estrangeira pra sua entidade. Por exemplo, uma classe Aluno que tenha um relacionamento 1 pra 1 com uma classe Pessoa. Caso você crie uma propriedade **PessoaId** na classe Aluno, o Entity Framework vai inferir PessoaId como sendo a chave estrangeira deste relacionamento.
Como você apontou, como alternativa, é possível indicar pro Entity Framework qual propriedade você quer que seja a chave primária ou estrangeira, tanto com Annotations quanto na Model Configuration do Contexto.

**Comment by marcus-silveira** [comment]
![image](https://github.com/balta-io/forum/assets/102529629/fc6191f1-174f-45a0-8087-0b39d2bc32b8)
![image](https://github.com/balta-io/forum/assets/102529629/d7b1bad7-b236-48fd-96ca-339ba46290c9)

**Reply by henriqueweege** [reply]
Fala, meu caro!
Então, seguindo [essa doc da Microsoft](https://learn.microsoft.com/en-us/ef/core/modeling/entity-properties?tabs=data-annotations%2Cwithout-nrt#included-and-excluded-properties) você pode ver que apenas as propriedades públicas que contém "get" **e** "set" são incluídas no modelo. Portanto, caso você não queira fazer o mapeamento utilizando Fluent API, você vai precisar corrigir suas models. 
Valeeeeu!

**Reply by malighieri** [reply]
Bem observado! Passei batido nas propriedades Id sem o SET!

**Comment by malighieri** [comment]
Bom dia! Obrigado por postar as classes. Pelo que olhei, as nomenclaturas estão de acordo com a nomenclatura que o Entity usaria pra inferir as chaves. Então eu não sei o que ocorreu pra ele não proceder assim. Inclusive na documentação de Code First da Microsoft que você mandou, explica estas convenções que falei.

De qualquer forma, eu costumo, até pra critério de documentação, informar na entidade as chaves. E isso pode ser feito de duas formas. A que mais gosto de usar é a que você usou inicialmente, que é chamada de Fluent API. Ou pode ser feita da forma que você fez por último com as Data Annotations.
